File: python/examples/dd_table_for_deal.py
#!/usr/bin/env python3
"""Print the double-dummy table and par for a deal from PBN on the CLI or a file.

Python counterpart to examples/dd_table_for_deal.cpp.
"""

from __future__ import annotations

import os
import re
import sys
from pathlib import Path

from dds3 import calc_all_tables_pbn, calc_par_from_table

PBN_FILE_MAX = 16 * 1024 * 1024  # safety cap for whole PBN exports
PBN_DEAL_MAX = 80  # matches DdTableDealPBN::cards

_CONTRACT_RE = re.compile(
    r"^(N|E|S|W|NS|EW)\s+(\d+)([SHDCN])(x)?$",
    re.IGNORECASE,
)

_VULNERABLE_ALIASES = {
    "none": 0,
    "0": 0,
    "both": 1,
    "1": 1,
    "ns": 2,
    "2": 2,
    "ew": 3,
    "3": 3,
}

# res_table[strain][hand]: strain 0-3 = S,H,D,C; 4 = NT. Columns match C++ print_table.
_STRAIN_ROWS = (("NT", 4), ("S", 0), ("H", 1), ("D", 2), ("C", 3))
_HAND_COLUMNS = (("North", 0), ("South", 2), ("East", 1), ("West", 3))

_DDS_FULL_LINE = 80
_DDS_HAND_OFFSET = 12
_DDS_HAND_LINES = 12

_BIT_MAP_RANK = [
    0x0000, 0x0000, 0x0001, 0x0002, 0x0004, 0x0008, 0x0010, 0x0020,
    0x0040, 0x0080, 0x0100, 0x0200, 0x0400, 0x0800, 0x1000, 0x2000,
]
_CARD_RANK_CHARS = "xx23456789TJQKA-"

_DEAL_TAG_RE = re.compile(r'\[Deal\s*"([^"]*)"', re.IGNORECASE)


def _read_pbn_stream(stream) -> str | None:
    data = stream.read(PBN_FILE_MAX + 1)
    if not data:
        return None
    text = data if isinstance(data, str) else data.decode("utf-8", errors="replace")
    if len(text) > PBN_FILE_MAX:
        raise ValueError(f"PBN input too large (max {PBN_FILE_MAX} characters)")
    return text


def _read_pbn_file(path: str) -> str | None:
    candidates = [Path(path)]
    workspace = os.environ.get("BUILD_WORKSPACE_DIRECTORY")
    if workspace is not None:
        candidates.append(Path(workspace) / path)
    for candidate in candidates:
        try:
            text = candidate.read_text(encoding="utf-8", errors="replace")
            if len(text) > PBN_FILE_MAX:
                raise ValueError(f"PBN file too large (max {PBN_FILE_MAX} characters)")
            return text
        except OSError:
            continue
    return None


def _extract_deal_tags(text: str) -> list[str]:
    return [match.group(1) for match in _DEAL_TAG_RE.finditer(text)]


def _unique_deals(deals: list[str]) -> list[str]:
    """Return deals in first-seen order with exact duplicates removed."""
    unique: list[str] = []
    seen: set[str] = set()
    for deal in deals:
        if deal in seen:
            continue
        seen.add(deal)
        unique.append(deal)
    return unique


def _looks_like_path(arg: str) -> bool:
    """True when arg is more likely a filename than a raw PBN deal string."""
    if "/" in arg or "\\" in arg:
        return True
    lower = arg.lower()
    return lower.endswith(".pbn") or lower.endswith(".txt")


def _load_deals(arg: str) -> list[str]:
    if arg == "-":
        text = _read_pbn_stream(sys.stdin)
        if text is None:
            raise ValueError("No PBN input on stdin")
        source = "stdin"
    else:
        text = _read_pbn_file(arg)
        source = arg if text is not None else None

    if source is not None:
        deals = _extract_deal_tags(text)
        if not deals:
            raise ValueError(f'No [Deal "..."] tag found in {source}')
        return deals

    if _looks_like_path(arg):
        raise ValueError(f"Cannot read file: {arg}")

    if len(arg) >= PBN_DEAL_MAX:
        raise ValueError(f"PBN deal too long (max {PBN_DEAL_MAX - 1} characters)")
    return [arg]


def _parse_vulnerable(text: str) -> int:
    try:
        return _VULNERABLE_ALIASES[text.lower()]
    except KeyError as exc:
        raise ValueError(
            "Invalid --vul value (use none|both|ns|ew or 0|1|2|3)"
        ) from exc


def _parse_limit(text: str) -> int:
    if not text.isdigit() or int(text) < 1:
        raise ValueError("Invalid --limit value (use a positive integer)")
    return int(text)


def _parse_cli(argv: list[str]) -> tuple[str, int, int | None] | None:
    """Return (deal_arg, vulnerable, limit) or None for help.

    limit is None when unrestricted. Raises ValueError on bad args.
    """
    deal: str | None = None
    vulnerable = 0
    limit: int | None = None
    i = 1
    while i < len(argv):
        arg = argv[i]
        if arg in ("-h", "--help"):
            return None
        if arg == "--vul":
            if i + 1 >= len(argv):
                raise ValueError("--vul requires a value (none|both|ns|ew or 0|1|2|3)")
            vulnerable = _parse_vulnerable(argv[i + 1])
            i += 2
            continue
        if arg == "--limit":
            if i + 1 >= len(argv):
                raise ValueError("--limit requires a positive integer")
            limit = _parse_limit(argv[i + 1])
            i += 2
            continue
        if arg.startswith("-") and arg != "-":
            raise ValueError(f"Unknown option: {arg}")
        if deal is not None:
            raise ValueError("Only one deal argument is allowed")
        deal = arg
        i += 1

    if deal is None:
        if not sys.stdin.isatty():
            deal = "-"
        else:
            raise ValueError("missing deal argument")

    return deal, vulnerable, limit


def _print_usage(prog: str) -> None:
    print(
        f"Usage: {prog} [--vul none|both|ns|ew|0|1|2|3] [--limit N] "
        f"<pbn_deal_or_file>\n"
        f"       {prog} -h | --help\n"
        "\n"
        "Calculate double-dummy tricks and par for all strains and leads.\n"
        "\n"
        "Arguments:\n"
        "  <pbn_deal_or_file>  DDS PBN deal string, or path to a .pbn file\n"
        "  --vul              Vulnerability: none|both|ns|ew or 0|1|2|3"
        " (default: none)\n"
        "  --limit            Solve only the first N unique deals\n"
        "\n"
        'If stdin is not a terminal, PBN is read from stdin (all [Deal "..."] tags).\n'
        "\n"
        "Examples:\n"
        f'  {prog} "N:73.QJT.AQ54.T752 QT6.876.KJ9.AQ84 '
        f'5.A95432.7632.K6 AKJ9842.K.T8.J93"\n'
        f"  {prog} --vul ns hands/example.pbn\n"
        f"  {prog} --limit 3 hands/multi_board.pbn\n"
        f"  {prog} < hands/example.pbn\n",
        file=sys.stderr,
    )


def _rawscore_undertricks(tricks: int, is_vul: bool) -> int:
    """Match library/src/par.cpp rawscore(-1, tricks, is_vul)."""
    if is_vul:
        return -300 * tricks + 100
    if tricks <= 3:
        return -200 * tricks + 100
    return -300 * tricks + 400


def _side_vulnerable(seats: str, vulnerable: int) -> bool:
    if vulnerable == 1:
        return True
    if seats in ("NS", "N", "S"):
        return vulnerable == 2
    if seats in ("EW", "E", "W"):
        return vulnerable == 3
    return False


def _sacrifice_undertricks(score: int, vulnerable: int, seats: str) -> int:
    """Infer sacrifice undertricks from the declaring-side par score."""
    is_vul = _side_vulnerable(seats, vulnerable)
    for tricks in range(1, 14):
        sacrifice_score = _rawscore_undertricks(tricks, is_vul)
        if sacrifice_score == score or -sacrifice_score == score:
            return tricks
    raise ValueError(f"cannot infer undertricks for par score {score}")


def _contract_body(side_contracts: str) -> str | None:
    if ":" not in side_contracts:
        return None
    body = side_contracts.split(":", 1)[1].strip()
    return body or None


def _normalize_contract_piece(piece: str) -> tuple[str, str, str] | None:
    """Expand DDS multi-level encodings like 'EW 45S' into ('EW', '4S', '+1')."""
    match = _CONTRACT_RE.match(piece.strip())
    if match is None:
        return None

    seats, levels, denom, doubled = match.groups()
    seats = seats.upper()
    denom = denom.upper()
    digits = [int(ch) for ch in levels]
    level = digits[0]
    over = digits[-1] - digits[0]

    contract = f"{level}{denom}{'x' if doubled else ''}"
    if doubled:
        return seats, contract, "="

    result = f"+{over}" if over > 0 else "="
    return seats, contract, result


def _normalize_par_body(body: str) -> tuple[str, str] | None:
    """Normalize comma-separated contracts; result comes from the first make."""
    pieces = [p.strip() for p in body.split(",") if p.strip()]
    if not pieces:
        return None

    normalized: list[str] = []
    result = "="
    for i, piece in enumerate(pieces):
        parsed = _normalize_contract_piece(piece)
        if parsed is None:
            return None
        seats, contract, piece_result = parsed
        if i == 0:
            normalized.append(f"{seats} {contract}")
            result = piece_result
        else:
            normalized.append(contract)
    return ", ".join(normalized), result


def _first_seats(body: str) -> str | None:
    match = _CONTRACT_RE.match(body.split(",", 1)[0].strip())
    if match is None:
        return None
    return match.group(1).upper()


def _declaring_score(par_results: dict, seats: str) -> int:
    scores = par_results["par_score"]
    ns_score = int(scores[0].split()[1])
    ew_score = int(scores[1].split()[1])
    if seats in ("NS", "N", "S"):
        return ns_score
    return ew_score


def _format_par_line(par_results: dict, *, vulnerable: int) -> str | None:
    """Return a one-line par summary for the declaring side."""
    scores = par_results["par_score"]
    contracts = par_results["par_contracts_string"]

    ns_score = int(scores[0].split()[1])
    ew_score = int(scores[1].split()[1])
    if ns_score == 0 and ew_score == 0:
        return "Par: 0"

    body = _contract_body(contracts[0])
    if body is None:
        return None

    seats = _first_seats(body)
    if seats is None:
        return None

    normalized = _normalize_par_body(body)
    if normalized is None:
        return None
    body, make_result = normalized

    score = _declaring_score(par_results, seats)
    is_sacrifice = "x" in body.lower()
    if is_sacrifice:
        result = f"-{_sacrifice_undertricks(score, vulnerable, seats)}"
    else:
        result = make_result

    return f"Par: {body} {result} {score}"


def _print_par_verbose(par_results: dict) -> None:
    """Match examples/hands.cpp print_par."""
    scores = par_results["par_score"]
    contracts = par_results["par_contracts_string"]
    print(f"NS score: {scores[0]}")
    print(f"EW score: {scores[1]}")
    print(f"NS list : {contracts[0]}")
    print(f"EW list : {contracts[1]}")
    print()


def _print_par(par_results: dict, *, vulnerable: int = 0) -> None:
    line = _format_par_line(par_results, vulnerable=vulnerable)
    if line is None:
        _print_par_verbose(par_results)
    else:
        print(line)


def _is_card(ch: str) -> int:
    ch = ch.upper()
    ranks = "23456789TJQKA"
    return ranks.index(ch) + 2 if ch in ranks else 0


def _convert_pbn(pbn_deal: str) -> list[list[int]]:
    """Match examples/hands.cpp convert_pbn (4 hands x 4 suits bitmasks)."""
    remain = [[0] * 4 for _ in range(4)]
    bp = 0
    while (
        bp < 3
        and bp < len(pbn_deal)
        and pbn_deal[bp] not in "NWESnwes"
    ):
        bp += 1
    if bp >= 3 or bp >= len(pbn_deal):
        return remain

    first = {"N": 0, "E": 1, "S": 2, "W": 3}[pbn_deal[bp].upper()]
    bp += 2
    hand_rel_first = 0
    suit_in_hand = 0

    while bp < 80 and bp < len(pbn_deal):
        ch = pbn_deal[bp]
        card = _is_card(ch)
        if card:
            if first == 0:
                hand = hand_rel_first
            elif first == 1:
                hand = 1 if hand_rel_first == 0 else 0 if hand_rel_first == 3 else hand_rel_first + 1
            elif first == 2:
                hand = 2 if hand_rel_first == 0 else 3 if hand_rel_first == 1 else hand_rel_first - 2
            else:
                hand = 3 if hand_rel_first == 0 else hand_rel_first - 1
            remain[hand][suit_in_hand] |= _BIT_MAP_RANK[card] << 2
        elif ch == ".":
            suit_in_hand += 1
        elif ch == " ":
            hand_rel_first += 1
            suit_in_hand = 0
        bp += 1
    return remain


def _print_pbn_hand(title: str, pbn_deal: str) -> None:
    """Match examples/hands.cpp print_pbn_hand / print_hand."""
    remain_cards = _convert_pbn(pbn_deal)
    text = [[" "] * _DDS_FULL_LINE for _ in range(_DDS_HAND_LINES)]
    row_ends = [_DDS_FULL_LINE] * _DDS_HAND_LINES

    for h in range(4):
        if h == 0:
            offset, line = _DDS_HAND_OFFSET, 0
        elif h == 1:
            offset, line = 2 * _DDS_HAND_OFFSET, 4
        elif h == 2:
            offset, line = _DDS_HAND_OFFSET, 8
        else:
            offset, line = 0, 4

        for s in range(4):
            row = line + s
            c = offset
            for r in range(14, 1, -1):
                if (remain_cards[h][s] >> 2) & _BIT_MAP_RANK[r]:
                    text[row][c] = _CARD_RANK_CHARS[r]
                    c += 1
            if c == offset:
                text[row][c] = "-"
                c += 1
            if h != 3:
                row_ends[row] = c

    sys.stdout.write(title)
    dash_len = max(0, len(title) - 1)
    print("-" * dash_len)
    for i in range(_DDS_HAND_LINES):
        print("".join(text[i][: row_ends[i]]))
    print()


def _print_table(res_table: list[list[int]]) -> None:
    """Match examples/hands.cpp print_table (%5s %-5s ... / %5c %5d ...)."""
    print(f"{'':>5} {'North':<5} {'South':<5} {'East':<5} {'West':<5}")

    _, nt_strain = _STRAIN_ROWS[0]
    print(
        f"{'NT':>5} "
        f"{res_table[nt_strain][_HAND_COLUMNS[0][1]]:5d} "
        f"{res_table[nt_strain][_HAND_COLUMNS[1][1]]:5d} "
        f"{res_table[nt_strain][_HAND_COLUMNS[2][1]]:5d} "
        f"{res_table[nt_strain][_HAND_COLUMNS[3][1]]:5d}"
    )

    for label, strain in _STRAIN_ROWS[1:]:
        print(
            f"{label:>5} "
            f"{res_table[strain][_HAND_COLUMNS[0][1]]:5d} "
            f"{res_table[strain][_HAND_COLUMNS[1][1]]:5d} "
            f"{res_table[strain][_HAND_COLUMNS[2][1]]:5d} "
            f"{res_table[strain][_HAND_COLUMNS[3][1]]:5d}"
        )
    print()


def main(argv: list[str] | None = None) -> int:
    argv = list(sys.argv if argv is None else argv)
    prog = Path(argv[0]).name

    try:
        parsed = _parse_cli(argv)
    except ValueError as exc:
        print(exc, file=sys.stderr)
        _print_usage(prog)
        return 1

    if parsed is None:
        _print_usage(prog)
        return 0

    input_arg, vulnerable, limit = parsed

    try:
        pbn_deals = _unique_deals(_load_deals(input_arg))
    except ValueError as exc:
        print(exc, file=sys.stderr)
        return 1

    if limit is not None:
        pbn_deals = pbn_deals[:limit]

    if any(len(deal) >= PBN_DEAL_MAX for deal in pbn_deals):
        print(
            f"PBN deal too long (max {PBN_DEAL_MAX - 1} characters)",
            file=sys.stderr,
        )
        return 1

    deal_count = len(pbn_deals)
    for deal_no, pbn_deal in enumerate(pbn_deals, start=1):
        try:
            result = calc_all_tables_pbn([pbn_deal])
        except (ValueError, RuntimeError) as exc:
            print(f"DDS error: {exc}", file=sys.stderr)
            return 1

        tables = result.get("tables")
        if not tables:
            print("DDS error: no table returned", file=sys.stderr)
            return 1

        table = tables[0]
        try:
            par_results = calc_par_from_table(table, vulnerable=vulnerable)
        except (ValueError, RuntimeError) as exc:
            print(f"DDS error: {exc}", file=sys.stderr)
            return 1

        title = (
            "dd_table_for_deal:\n"
            if deal_count == 1
            else f"Deal {deal_no}:\n"
        )
        _print_pbn_hand(title, pbn_deal)
        _print_table(table["res_table"])
        _print_par(par_results, vulnerable=vulnerable)
        if deal_count > 1:
            print()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/dd_table_for_deal.cpp
/*
   DDS, a bridge double dummy solver.

   Copyright (C) 2006-2014 by Bo Haglund /
   2014-2016 by Bo Haglund & Soren Hein.

   See LICENSE and README.
*/


// Print the double-dummy table and par for a deal from the command line or a PBN file.

// Coded by Cursor, based on calc_dd_table.cpp

#include <algorithm>
#include <cctype>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <optional>
#include <regex>
#include <string>
#include <string_view>
#include <vector>
#if defined(_WIN32)
#include <io.h>
#else
#include <unistd.h>
#endif
#include <api/dll.h>
#include "hands.hpp"


namespace {

constexpr std::size_t PBN_FILE_MAX = 16 * 1024 * 1024;
constexpr std::size_t PBN_DEAL_MAX = sizeof(DdTableDealPBN::cards);

const std::regex DEAL_TAG_RE{
    R"re(\[Deal\s*"([^"]*)")re",
    std::regex::icase};


static auto stdin_is_tty() -> bool
{
#if defined(_WIN32)
  return _isatty(_fileno(stdin)) != 0;
#else
  return isatty(STDIN_FILENO) != 0;
#endif
}


auto read_pbn_stream(std::istream& in) -> std::optional<std::string>
{
  std::string text;
  text.reserve(64 * 1024);
  char buffer[4096];
  while (in)
  {
    in.read(buffer, static_cast<std::streamsize>(sizeof(buffer)));
    const auto n = in.gcount();
    if (n > 0)
      text.append(buffer, static_cast<std::size_t>(n));
    if (text.size() > PBN_FILE_MAX)
    {
      std::cerr << "PBN input too large (max " << PBN_FILE_MAX << " characters)\n";
      return std::nullopt;
    }
  }
  if (text.empty())
    return std::nullopt;
  return text;
}


auto read_pbn_file(const std::filesystem::path& path) -> std::optional<std::string>
{
  std::ifstream file(path, std::ios::binary);
  if (!file)
  {
    return std::nullopt;
  }

  return read_pbn_stream(file);
}


auto read_pbn_file_workspace_relative(std::string_view path)
    -> std::optional<std::string>
{
  if (auto text = read_pbn_file(std::filesystem::path(path)))
  {
    return text;
  }

  // bazelisk run uses a runfiles cwd; BUILD_WORKSPACE_DIRECTORY is the repo root.
  if (const char* workspace = std::getenv("BUILD_WORKSPACE_DIRECTORY"))
  {
    return read_pbn_file(std::filesystem::path(workspace) / path);
  }

  return std::nullopt;
}


auto extract_deal_tags(std::string_view text) -> std::vector<std::string>
{
  std::vector<std::string> deals;
  auto begin = text.cbegin();
  const auto end = text.cend();
  std::match_results<std::string_view::const_iterator> match;
  while (std::regex_search(begin, end, match, DEAL_TAG_RE) && match.size() > 1)
  {
    deals.emplace_back(match[1].first, match[1].second);
    begin = match[0].second;
  }
  return deals;
}


auto parse_vulnerable(std::string_view text) -> std::optional<int>
{
  std::string lower(text);
  std::transform(lower.begin(), lower.end(), lower.begin(),
                 [](unsigned char c) { return static_cast<char>(std::tolower(c)); });

  if (lower == "none" || lower == "0")
    return 0;
  if (lower == "both" || lower == "1")
    return 1;
  if (lower == "ns" || lower == "2")
    return 2;
  if (lower == "ew" || lower == "3")
    return 3;
  return std::nullopt;
}


auto looks_like_path(std::string_view arg) -> bool
{
  if (arg.find('/') != std::string_view::npos
      || arg.find('\\') != std::string_view::npos)
  {
    return true;
  }

  if (arg.size() >= 4)
  {
    std::string lower(arg.substr(arg.size() - 4));
    std::transform(lower.begin(), lower.end(), lower.begin(),
                   [](unsigned char c) {
                     return static_cast<char>(std::tolower(c));
                   });
    if (lower == ".pbn" || lower == ".txt")
      return true;
  }
  return false;
}


auto load_deals(std::string_view arg) -> std::optional<std::vector<std::string>>
{
  if (arg == "-")
  {
    const auto text = read_pbn_stream(std::cin);
    if (!text)
    {
      std::cerr << "No PBN input on stdin\n";
      return std::nullopt;
    }

    const auto deals = extract_deal_tags(*text);
    if (deals.empty())
    {
      std::cerr << "No [Deal \"...\"] tag found in stdin\n";
      return std::nullopt;
    }

    return deals;
  }

  if (const auto text = read_pbn_file_workspace_relative(arg))
  {
    const auto deals = extract_deal_tags(*text);
    if (deals.empty())
    {
      std::cerr << "No [Deal \"...\"] tag found in " << arg << "\n";
      return std::nullopt;
    }

    return deals;
  }

  if (looks_like_path(arg))
  {
    std::cerr << "Cannot read file: " << arg << "\n";
    return std::nullopt;
  }

  if (arg.size() >= PBN_DEAL_MAX)
  {
    std::cerr << "PBN deal too long (max " << (PBN_DEAL_MAX - 1)
              << " characters)\n";
    return std::nullopt;
  }

  return std::vector<std::string>{std::string(arg)};
}


auto denom_char(int denom) -> char
{
  static constexpr char kDenomChars[] = "NSHDC";
  if (denom < 0 || denom > 4)
    return '?';
  return kDenomChars[denom];
}


auto seat_name(int seats) -> const char *
{
  static const char * kSeatNames[] = {"N", "E", "S", "W", "NS", "EW"};
  if (seats < 0 || seats > 5)
    return "?";
  return kSeatNames[seats];
}


auto format_par_line(
    ParResultsMaster const sidesRes[2],
    int vulnerable,
    char * line,
    std::size_t line_size) -> bool
{
  if (line_size < 32)
    return false;

  if (sidesRes[0].score == 0 && sidesRes[1].score == 0)
  {
    std::snprintf(line, line_size, "Par: 0");
    return true;
  }

  if (sidesRes[0].number != 1 || sidesRes[1].number != 1)
    return false;

  const ContractType& contract = sidesRes[0].contracts[0];
  const char * seats = seat_name(contract.seats);
  const char result_sign = contract.under_tricks > 0 ? '-' : '+';
  const int result_value =
      contract.under_tricks > 0 ? contract.under_tricks : contract.over_tricks;

  if (contract.under_tricks > 0)
  {
    std::snprintf(
        line,
        line_size,
        "Par: %s %d%c%c %c%d %d",
        seats,
        contract.level,
        denom_char(contract.denom),
        'x',
        result_sign,
        result_value,
        sidesRes[0].score);
  }
  else
  {
    std::snprintf(
        line,
        line_size,
        "Par: %s %d%c %c%d %d",
        seats,
        contract.level,
        denom_char(contract.denom),
        result_sign,
        result_value,
        sidesRes[0].score);
  }

  return true;
}


auto print_par_or_verbose(
    DdTableResults const * table,
    int vulnerable) -> void
{
  ParResultsMaster sidesRes[2];
  const int res = SidesParBin(table, sidesRes, vulnerable);
  if (res != RETURN_NO_FAULT)
  {
    char line[80];
    ErrorMessage(res, line);
    fprintf(stderr, "DDS error: %s\n", line);
    return;
  }

  char line[80];
  if (format_par_line(sidesRes, vulnerable, line, sizeof(line)))
  {
    printf("%s\n", line);
    return;
  }

  ParResults par;
  const int par_res = Par(table, &par, vulnerable);
  if (par_res != RETURN_NO_FAULT)
  {
    ErrorMessage(par_res, line);
    fprintf(stderr, "DDS error: %s\n", line);
    return;
  }

  print_par(&par);
}


auto process_deal(
    std::string const& deal,
    std::size_t deal_no,
    std::size_t deal_count,
    int vulnerable) -> bool
{
  DdTableDealPBN tableDealPBN{};
  if (deal.size() >= sizeof(tableDealPBN.cards))
  {
    fprintf(stderr,
            "PBN deal too long (max %zu characters)\n",
            sizeof(tableDealPBN.cards) - 1);
    return false;
  }

  std::copy_n(deal.begin(), deal.size(), tableDealPBN.cards);
  tableDealPBN.cards[deal.size()] = '\0';

  DdTableResults table;
  char line[80];

  const int res = CalcDDtablePBN(tableDealPBN, &table);
  if (res != RETURN_NO_FAULT)
  {
    ErrorMessage(res, line);
    fprintf(stderr, "DDS error: %s\n", line);
    return false;
  }

  if (deal_count == 1)
    sprintf(line, "dd_table_for_deal:\n");
  else
    sprintf(line, "Deal %zu:\n", deal_no);

  print_pbn_hand(line, tableDealPBN.cards);
  print_table(&table);
  print_par_or_verbose(&table, vulnerable);
  if (deal_count > 1)
    printf("\n");
  return true;
}

}  // namespace


static auto print_usage(const char * prog) -> void
{
  fprintf(stderr,
          "Usage: %s [--vul none|both|ns|ew] <pbn_deal_or_file>\n"
          "       %s -h | --help\n"
          "\n"
          "Calculate double-dummy tricks and par for all strains and leads.\n"
          "\n"
          "Arguments:\n"
          "  <pbn_deal_or_file>  DDS PBN deal string, or path to a .pbn file\n"
          "  --vul              Vulnerability: none, both, ns, ew (default: none)\n"
          "\n"
          "If stdin is not a terminal, PBN is read from stdin (all [Deal \"...\"] tags).\n"
          "\n"
          "Examples:\n"
          "  %s \"N:73.QJT.AQ54.T752 QT6.876.KJ9.AQ84 "
          "5.A95432.7632.K6 AKJ9842.K.T8.J93\"\n"
          "  %s --vul ns hands/example.pbn\n"
          "  %s < hands/example.pbn\n",
          prog,
          prog,
          prog,
          prog,
          prog);
}


auto main(int argc, char * argv[]) -> int
{
  const char * input = nullptr;
  int vulnerable = 0;

  for (int i = 1; i < argc; ++i)
  {
    if (strcmp(argv[i], "-h") == 0 || strcmp(argv[i], "--help") == 0)
    {
      print_usage(argv[0]);
      return 0;
    }
    if (strcmp(argv[i], "--vul") == 0)
    {
      if (i + 1 >= argc)
      {
        fprintf(stderr, "--vul requires a value (none|both|ns|ew or 0|1|2|3)\n");
        print_usage(argv[0]);
        return 1;
      }
      const auto vul = parse_vulnerable(argv[++i]);
      if (!vul)
      {
        fprintf(stderr, "Invalid --vul value (use none|both|ns|ew or 0|1|2|3)\n");
        print_usage(argv[0]);
        return 1;
      }
      vulnerable = *vul;
      continue;
    }
    if (argv[i][0] == '-' && strcmp(argv[i], "-") != 0)
    {
      fprintf(stderr, "Unknown option: %s\n", argv[i]);
      print_usage(argv[0]);
      return 1;
    }
    if (input != nullptr)
    {
      print_usage(argv[0]);
      return 1;
    }
    input = argv[i];
  }

  if (input == nullptr)
  {
    if (!stdin_is_tty())
      input = "-";
    else
    {
      print_usage(argv[0]);
      return 1;
    }
  }

  const auto deals = load_deals(input);
  if (!deals)
  {
    return 1;
  }

  for (std::size_t i = 0; i < deals->size(); ++i)
  {
    if (!process_deal((*deals)[i], i + 1, deals->size(), vulnerable))
      return 1;
  }

  return 0;
}
